src/mlpack/methods/reinforcement_learning/environment/maze.hpp

```diff
@@ -0,0 +1,406 @@
+   */
+  Maze(const size_t rows = 4,
+        const size_t columns = 4,
+        const size_t maxSteps = 200) :
+      maxSteps(maxSteps),
+      stepsPerformed(0)
+  { 
+    MazeGeneration(rows, columns);
+    
+    directions = std::unordered_map<Action::actions, arma::vec>({
+            { Action::actions::left,  arma::vec({0, -1}) },
+            { Action::actions::right,  arma::vec({0, 1}) },
+            { Action::actions::up,  arma::vec({-1, 0}) },
+            { Action::actions::down,  arma::vec({1, 0}) }
+        });
+
+  }
+
+  /**
+   * Generate Maze using Random Iterative DFS Algorithm with 
+   * 0.4 * rows * column steps
+   * @param rows number of rows in maze.
+   * @param columns number of columns in maze.
+   * 
+   */
+  void MazeGeneration(const size_t rows, const size_t columns)
+  {
+    maze = arma::mat(rows,columns);
+    arma::umat visited;
+    visited.zeros(rows, columns);
+
+    std::vector<arma::ivec> moves = {arma::ivec({0, -1}), arma::ivec({0, 1}), arma::ivec({-1, 0}), arma::ivec({1, 0}) };
+
+    const size_t maxNumberOfMazeSteps = 2 * std::sqrt(rows * columns);
+
+    std::stack<arma::ivec> cellsInPath;
+
+    arma::ivec startingPoint = {arma::randi(arma::distr_param(0, rows-1)), arma::randi(arma::distr_param(0, columns-1))};
+
+    cellsInPath.push(startingPoint);
+
+    size_t numberOfMazeSteps = 0;
+ 
+    // loop till stack is empty
+    while (!cellsInPath.empty())
+    {
+        // Pop a vertex from the stack
+        arma::ivec lastCell = cellsInPath.top();
+        cellsInPath.pop();
+
+        ++numberOfMazeSteps;
+
+        // Break for Max number of maze steps.
+        if(numberOfMazeSteps > maxNumberOfMazeSteps)
+          break;
+ 
+        // Mark the particular row and column as visited.
+        visited(lastCell(0),lastCell(1)) = 1;
+
+        // If visitable nodes are present from that node.
+        bool visitableNodes = false;
+
+        // Randomly choose a direction and check if the adjacent node in 
+        // that direction is visited or not.
+        for (size_t counter = 0; counter < 10 ; ++counter)
+        {    
+            size_t index = arma::randi(arma::distr_param(0, 3));
+            arma::ivec nextCell = arma::ivec({lastCell(0) + moves[index](0), {lastCell(1) + moves[index](1)}});
+            
+            // Check if particular row and column index are out of bounds of maze.
+            bool outOfBounds = nextCell(0)>= maze.n_rows || nextCell(1)>= maze.n_cols
+        || nextCell(0) < 0 || nextCell(1) < 0 ;
+            if (!outOfBounds && !visited(nextCell(0),nextCell(1))) 
+            {
+                cellsInPath.push(lastCell);
+                cellsInPath.push(nextCell);
+                break;
+            }
+        }
+    }
+
+    // Pop last ({row,column}) from the stack
+      arma::ivec goalCell = cellsInPath.top();
+      cellsInPath.pop();
+
+    // Set the ({row,column}) asa goal cell
+      maze(goalCell(0),goalCell(1)) = 1;
```

Comment:
Looks like the indentation is off here, some of the lines below as well, see https://github.com/mlpack/mlpack/wiki/DesignGuidelines#line-length-and-wrapping for more details.